## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [39]:
import pandas as pd

df = pd.read_csv('/Users/jedmurph/Documents/CODE/pandas-challenge-1/Starter_Code/Resources/client_dataset.csv')
df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year         item_id     category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   
6   8080280  2023-04-19          16        2023  UAC50632-25-4E  consumables   
7   8946782  2023-02-14           7        2023  UEU25316-01-1U  consumables   
8   9394118  2023-05-13          19        2023  DDA06248-98-7Z  consumables   
9   9665852  2023-04-14          15        2023  DUU85042-87-5E    furniture   

          subcategory  unit_price  unit_cost  unit_weight  qty  line_number  
0            wall art     1096.80     762.71         7.50  105            1  
1                pens       24.95      15.09         1.49   21            0  
2  project management       13.52       7.86         1.68   39            6  
3                pens       36.42      24.85         1.23   29            3  
4                misc      195.10     108.17        46.43   20            1  
5              plants      172.84     154.44         2.79   78            0  
6    kitchen supplies      245.56     231.95         1.73   56            5  
7                pens       20.99      16.97         5.12   22            4  
8      paper products       83.86      59.00         0.27   49            3  
9             seating       66.88      54.15         2.47   78            1

In [40]:
# View the column names in the data
columns_list = df.columns

print(columns_list)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [41]:
# Use the describe function to gather some basic statistics

df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [42]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.count(axis='columns')


0        18
1        18
2        18
3        18
4        18
         ..
54634    18
54635    18
54636    18
54637    18
54638    18
Length: 54639, dtype: int64

In [43]:
df.isna().sum()

first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64

In [44]:
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [45]:
df["last"].unique()

array(['Harding', 'Myers', 'Watson', 'Baker', 'Schwartz', 'Reyes',
       'Moore', 'Wright', 'Brown', 'Garcia', 'Williams', 'Howard',
       'Gould', 'Taylor', 'Orozco', 'Jordan', 'Green', 'Stephenson',
       'Anderson', 'Roberts', 'Rhodes', 'Mendoza', 'Lowe', 'Rice', 'Cook',
       'Kennedy', 'Harris', 'Palmer', 'Robertson', 'Barnes', 'Wells',
       'Spencer', 'Brooks', 'Mitchell', 'Mccann', 'Bean', 'Griffith',
       'Wagner', 'Johnson', 'Newman', 'Hernandez', 'Davis', 'Simmons',
       'Turner', 'Aguilar', 'Kelly', 'Chase', 'Hobbs', 'Wilkinson',
       'Peterson', 'Cervantes', 'Carlson', 'Thompson', 'Vasquez', 'Ward',
       'Middleton', 'Arias', 'Ramirez', 'Pennington', 'Ellis', 'Mcmillan',
       'Morgan', 'Austin', 'Vincent', 'Campbell', 'Richardson', 'Colon',
       'Jackson', 'Espinoza', 'West', 'Schultz', 'Cannon', 'Hanson',
       'Gilbert', 'Powell', 'Lewis', 'Bennett', 'Evans', 'Martin',
       'Gross', 'Mullen', 'Hughes', 'Short', 'Castillo', 'Carr',
       'Farrell', 'C

In [46]:
df.value_counts(ascending=False)

first    last       job                             phone                   email                              client_id  order_id  order_date  order_week  order_year  item_id         category     subcategory       unit_price  unit_cost  unit_weight  qty  line_number
Aaron    Martinez   Armed forces technical officer  283.652.2505x6088       aaron.martinez@morton.biz          19820      2178609   2023-05-01  18          2023        ABX14990-37-1Z  furniture    tables            0.02        0.01       4.22         29   1              1
Melissa  Smith      Pilot, airline                  001-638-231-7128x99395  smithmelissa@miller.com            99231      2592231   2023-05-21  20          2023        XUB12331-18-1A  consumables  misc              30.40       25.88      2.00         50   4              1
                                                                                                                          3880150   2023-04-30  17          2023        BAB21028-35-2Z  co

In [47]:
total = df.sum()

print(total)

first          DonaldTiffanyShannonNathanChristinaJessicaBria...
last           HardingMyersWatsonBakerSchwartzReyesMooreWrigh...
job            ImmunologistMusic therapistImmunologistAccount...
phone          793-904-7725x39308201.442.4543x942687.737.9424...
email          harding.donald.7185@sullivan.commyers.t.6537@f...
client_id                                             2996286347
order_id                                            298885713037
order_date     2023-04-282023-05-192023-01-292023-04-252023-0...
order_week                                                620652
order_year                                             110534318
item_id        EUD29711-63-6UXDA18116-89-4AABE59463-05-7EZMM0...
category       decorconsumablessoftwareconsumablesconsumables...
subcategory    wall artpensproject managementpensmiscplantski...
unit_price                                            7445503.92
unit_cost                                             5433633.97
unit_weight              

In [48]:
# What three item categories had the most entries?
category = "category"
highest_three_values_in_column = df[category].value_counts().nlargest(3)

print(highest_three_values_in_column)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [49]:
# For the category with the most entries,
# which subcategory had the most entries?
subcategory = "subcategory"


most_common_subcategory = df[subcategory].value_counts().idxmax()

print(most_common_subcategory)


bathroom supplies


In [50]:
# Which five clients had the most entries in the data?
client_id = "client_id"

top_five_clients = df[client_id].value_counts().nlargest(5)
print(top_five_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [51]:
# Store the client ids of those top 5 clients in a list.

top_five_clients_ids = top_five_clients.index.tolist()
print(top_five_clients_ids)

[33615, 66037, 46820, 38378, 24741]


In [52]:
# How many total units (the qty column) did the
# client with the most entries order order?
top_client_id = df['client_id'].value_counts().idxmax()

total_units_ordered = df[df['client_id'] == top_client_id]['qty'].sum()

print(total_units_ordered)


64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [53]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df["subtotal"] = df["unit_price"] * df["qty"]



df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year         item_id     category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   
6   8080280  2023-04-19          16        2023  UAC50632-25-4E  consumables   
7   8946782  2023-02-14           7        2023  UEU25316-01-1U  consumables   
8   9394118  2023-05-13          19        2023  DDA06248-98-7Z  consumables   
9   9665852  2023-04-14          15        2023  DUU85042-87-5E    furniture   

          subcategory  unit_price  unit_cost  unit_weight  qty  line_number  \
0            wall art     1096.80     762.71         7.50  105            1   
1                pens       24.95      15.09         1.49   21            0   
2  project management       13.52       7.86         1.68   39            6   
3                pens       36.42      24.85         1.23   29            3   
4                misc      195.10     108.17        46.43   20            1   
5              plants      172.84     154.44         2.79   78            0   
6    kitchen supplies      245.56     231.95         1.73   56            5   
7                pens       20.99      16.97         5.12   22            4   
8      paper products       83.86      59.00         0.27   49            3   
9             seating       66.88      54.15         2.47   78            1   

    subtotal  
0  115164.00  
1     523.95  
2     527.28  
3    1056.18  
4    3902.00  
5   13481.52  
6   13751.36  
7     461.78  
8    4109.14  
9    5216.64

In [54]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.


df['shipping_price'] = df['unit_weight'] * df['qty'].apply(lambda x: 
                                          7 if x > 50 
                                          else 10)

df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year         item_id     category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   
6   8080280  2023-04-19          16        2023  UAC50632-25-4E  consumables   
7   8946782  2023-02-14           7        2023  UEU25316-01-1U  consumables   
8   9394118  2023-05-13          19        2023  DDA06248-98-7Z  consumables   
9   9665852  2023-04-14          15        2023  DUU85042-87-5E    furniture   

          subcategory  unit_price  unit_cost  unit_weight  qty  line_number  \
0            wall art     1096.80     762.71         7.50  105            1   
1                pens       24.95      15.09         1.49   21            0   
2  project management       13.52       7.86         1.68   39            6   
3                pens       36.42      24.85         1.23   29            3   
4                misc      195.10     108.17        46.43   20            1   
5              plants      172.84     154.44         2.79   78            0   
6    kitchen supplies      245.56     231.95         1.73   56            5   
7                pens       20.99      16.97         5.12   22            4   
8      paper products       83.86      59.00         0.27   49            3   
9             seating       66.88      54.15         2.47   78            1   

    subtotal  shipping_price  
0  115164.00           52.50  
1     523.95           14.90  
2     527.28           16.80  
3    1056.18           12.30  
4    3902.00          464.30  
5   13481.52           19.53  
6   13751.36           12.11  
7     461.78           51.20  
8    4109.14            2.70  
9    5216.64           17.29

In [55]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

df["total_price"] = df["subtotal"] + df["shipping_price"] + (df["subtotal"] + df["shipping_price"]) * 0.0925

df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year  ...     category  \
0   8953482  2023-04-28          17        2023  ...        decor   
1   8069089  2023-05-19          20        2023  ...  consumables   
2   1902144  2023-01-29           4        2023  ...     software   
3   9031802  2023-04-25          17        2023  ...  consumables   
4   1322274  2023-05-28          21        2023  ...  consumables   
5   1327730  2023-05-02          18        2023  ...        decor   
6   8080280  2023-04-19          16        2023  ...  consumables   
7   8946782  2023-02-14           7        2023  ...  consumables   
8   9394118  2023-05-13          19        2023  ...  consumables   
9   9665852  2023-04-14          15        2023  ...    furniture   

          subcategory unit_price  unit_cost  unit_weight  qty  line_number  \
0            wall art    1096.80     762.71         7.50  105            1   
1                pens      24.95      15.09         1.49   21            0   
2  project management      13.52       7.86         1.68   39            6   
3                pens      36.42      24.85         1.23   29            3   
4                misc     195.10     108.17        46.43   20            1   
5              plants     172.84     154.44         2.79   78            0   
6    kitchen supplies     245.56     231.95         1.73   56            5   
7                pens      20.99      16.97         5.12   22            4   
8      paper products      83.86      59.00         0.27   49            3   
9             seating      66.88      54.15         2.47   78            1   

    subtotal  shipping_price    total_price  
0  115164.00           52.50  125874.026250  
1     523.95           14.90     588.693625  
2     527.28           16.80     594.407400  
3    1056.18           12.30    1167.314400  
4    3902.00          464.30    4770.182750  
5   13481.52           19.53   14749.897125  
6   13751.36           12.11   15036.590975  
7     461.78           51.20     560.430650  
8    4109.14            2.70    4492.185200  
9    5216.64           17.29    5718.068525  

[10 rows x 21 columns]

In [56]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['line_cost'] = (df['unit_cost'] * df['qty']) + (df['shipping_price'] * df['qty'])

df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year  ...         subcategory  \
0   8953482  2023-04-28          17        2023  ...            wall art   
1   8069089  2023-05-19          20        2023  ...                pens   
2   1902144  2023-01-29           4        2023  ...  project management   
3   9031802  2023-04-25          17        2023  ...                pens   
4   1322274  2023-05-28          21        2023  ...                misc   
5   1327730  2023-05-02          18        2023  ...              plants   
6   8080280  2023-04-19          16        2023  ...    kitchen supplies   
7   8946782  2023-02-14           7        2023  ...                pens   
8   9394118  2023-05-13          19        2023  ...      paper products   
9   9665852  2023-04-14          15        2023  ...             seating   

  unit_price unit_cost  unit_weight  qty  line_number   subtotal  \
0    1096.80    762.71         7.50  105            1  115164.00   
1      24.95     15.09         1.49   21            0     523.95   
2      13.52      7.86         1.68   39            6     527.28   
3      36.42     24.85         1.23   29            3    1056.18   
4     195.10    108.17        46.43   20            1    3902.00   
5     172.84    154.44         2.79   78            0   13481.52   
6     245.56    231.95         1.73   56            5   13751.36   
7      20.99     16.97         5.12   22            4     461.78   
8      83.86     59.00         0.27   49            3    4109.14   
9      66.88     54.15         2.47   78            1    5216.64   

   shipping_price    total_price  line_cost  
0           52.50  125874.026250   85597.05  
1           14.90     588.693625     629.79  
2           16.80     594.407400     961.74  
3           12.30    1167.314400    1077.35  
4          464.30    4770.182750   11449.40  
5           19.53   14749.897125   13569.66  
6           12.11   15036.590975   13667.36  
7           51.20     560.430650    1499.74  
8            2.70    4492.185200    3023.30  
9           17.29    5718.068525    5572.32  

[10 rows x 22 columns]

In [57]:
# Create a column for the profit of
# each line using line cost and line price
df["line_profit"] = df["total_price"] - df["line_cost"]

df.head(10)


first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year  ... unit_price unit_cost  \
0   8953482  2023-04-28          17        2023  ...    1096.80    762.71   
1   8069089  2023-05-19          20        2023  ...      24.95     15.09   
2   1902144  2023-01-29           4        2023  ...      13.52      7.86   
3   9031802  2023-04-25          17        2023  ...      36.42     24.85   
4   1322274  2023-05-28          21        2023  ...     195.10    108.17   
5   1327730  2023-05-02          18        2023  ...     172.84    154.44   
6   8080280  2023-04-19          16        2023  ...     245.56    231.95   
7   8946782  2023-02-14           7        2023  ...      20.99     16.97   
8   9394118  2023-05-13          19        2023  ...      83.86     59.00   
9   9665852  2023-04-14          15        2023  ...      66.88     54.15   

  unit_weight  qty  line_number   subtotal  shipping_price    total_price  \
0        7.50  105            1  115164.00           52.50  125874.026250   
1        1.49   21            0     523.95           14.90     588.693625   
2        1.68   39            6     527.28           16.80     594.407400   
3        1.23   29            3    1056.18           12.30    1167.314400   
4       46.43   20            1    3902.00          464.30    4770.182750   
5        2.79   78            0   13481.52           19.53   14749.897125   
6        1.73   56            5   13751.36           12.11   15036.590975   
7        5.12   22            4     461.78           51.20     560.430650   
8        0.27   49            3    4109.14            2.70    4492.185200   
9        2.47   78            1    5216.64           17.29    5718.068525   

   line_cost   line_profit  
0   85597.05  40276.976250  
1     629.79    -41.096375  
2     961.74   -367.332600  
3    1077.35     89.964400  
4   11449.40  -6679.217250  
5   13569.66   1180.237125  
6   13667.36   1369.230975  
7    1499.74   -939.309350  
8    3023.30   1468.885200  
9    5572.32    145.748525  

[10 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [58]:
# Check your work using the totals above
total_price_per_order = df.groupby("order_id")["total_price"].sum()

order_id_list = [2742071, 2173913, 6128929]
for order_id in order_id_list:
    total_price = total_price_per_order.get(order_id, 0)

    print(f"Total price for order_id {order_id}: ${total_price:.2f}")

Total price for order_id 2742071: $139801.71
Total price for order_id 2173913: $119793.39
Total price for order_id 6128929: $752665.99


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [59]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
df["total_spent"] = df["qty"] * df["unit_price"]
client_spending = df.groupby("client_id")["total_spent"].sum()


for client_id in top_five_clients.index:
    client_total_spending = client_spending.loc[client_id]
    print(f"Client ID {client_id} spent: ${client_total_spending:.2f}")

   

Client ID 33615 spent: $5839032.11
Client ID 66037 spent: $7995708.38
Client ID 46820 spent: $7317356.03
Client ID 38378 spent: $8384321.15
Client ID 24741 spent: $70176885.25


In [65]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

df["total_units_purchased"] = df["qty"].groupby(df["client_id"]).sum()
df["total_shipping_price"] = df["shipping_price"].groupby(df["client_id"]).sum()
df["total_revenue"] = df["total_spent"].groupby(df["client_id"]).sum()
df["total_profit"] = df["total_revenue"] - df["total_shipping_price"]

top_five_clients_summary = df.groupby("client_id").agg(
    total_units_purchased=("total_units_purchased", "sum"),
    total_shipping_price=("total_shipping_price", "sum"),
    total_revenue=("total_revenue", "sum"),
    total_profit=("total_profit", "sum")
).sort_values(by="total_profit", ascending=False).head(5)

df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year  ...   subtotal  \
0   8953482  2023-04-28          17        2023  ...  115164.00   
1   8069089  2023-05-19          20        2023  ...     523.95   
2   1902144  2023-01-29           4        2023  ...     527.28   
3   9031802  2023-04-25          17        2023  ...    1056.18   
4   1322274  2023-05-28          21        2023  ...    3902.00   
5   1327730  2023-05-02          18        2023  ...   13481.52   
6   8080280  2023-04-19          16        2023  ...   13751.36   
7   8946782  2023-02-14           7        2023  ...     461.78   
8   9394118  2023-05-13          19        2023  ...    4109.14   
9   9665852  2023-04-14          15        2023  ...    5216.64   

  shipping_price    total_price  line_cost   line_profit  total_spent  \
0          52.50  125874.026250   85597.05  40276.976250    115164.00   
1          14.90     588.693625     629.79    -41.096375       523.95   
2          16.80     594.407400     961.74   -367.332600       527.28   
3          12.30    1167.314400    1077.35     89.964400      1056.18   
4         464.30    4770.182750   11449.40  -6679.217250      3902.00   
5          19.53   14749.897125   13569.66   1180.237125     13481.52   
6          12.11   15036.590975   13667.36   1369.230975     13751.36   
7          51.20     560.430650    1499.74   -939.309350       461.78   
8           2.70    4492.185200    3023.30   1468.885200      4109.14   
9          17.29    5718.068525    5572.32    145.748525      5216.64   

   total_units_purchased  total_shipping_price  total_revenue  total_profit  
0                    NaN                   NaN            NaN           NaN  
1                    NaN                   NaN            NaN           NaN  
2                    NaN                   NaN            NaN           NaN  
3                    NaN                   NaN            NaN           NaN  
4                    NaN                   NaN            NaN           NaN  
5                    NaN                   NaN            NaN           NaN  
6                    NaN                   NaN            NaN           NaN  
7                    NaN                   NaN            NaN           NaN  
8                    NaN                   NaN            NaN           NaN  
9                    NaN                   NaN            NaN           NaN  

In [78]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

columns_list_renamed = columns_list.rename(columns={
    "client_id": "Client ID",
    "order_id": "Order ID",
    "order_date": "Order Date",
    "order_week": "Order Week",
    "order_year": "Order Year",
    "subtotal": "Subtotal",
    "total_units_purchased": "Total Units",
    "total_shipping_price": "Total Shipping Price",
    "total_revenue": "Total Revenue",
    "total_profit": "Total Profit"
})

columns_list_renamed["Total Revenue"] = columns_list_renamed["Total Revenue"] / 1000000
columns_list_renamed["Total Profit"] = columns_list_renamed["Total Profit"] / 1000000
columns_list_renamed["Total Shipping Price"] = columns_list_renamed["Total Shipping Price"] / 1000000
df.head(5)

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...   subtotal shipping_price    total_price  \
0          17        2023  ...  115164.00           52.5  125874.026250   
1          20        2023  ...     523.95           14.9     588.693625   
2           4        2023  ...     527.28           16.8     594.407400   
3          17        2023  ...    1056.18           12.3    1167.314400   
4          21        2023  ...    3902.00          464.3    4770.182750   

   line_cost   line_profit  total_spent  total_units_purchased  \
0   85597.05  40276.976250    115164.00                    NaN   
1     629.79    -41.096375       523.95                    NaN   
2     961.74   -367.332600       527.28                    NaN   
3    1077.35     89.964400      1056.18                    NaN   
4   11449.40  -6679.217250      3902.00                    NaN   

   total_shipping_price  total_revenue  total_profit  
0                   NaN            NaN           NaN  
1                   NaN            NaN           NaN  
2                   NaN            NaN           NaN  
3                   NaN            NaN           NaN  
4                   NaN            NaN           NaN  

[5 rows x 28 columns]

In [ ]:
# Sort the updated data by "Total Profit" form highest to lowest


Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M

In [79]:
"Total Profit" = False

SyntaxError: cannot assign to literal here. Maybe you meant '==' instead of '='? (795694794.py, line 1)